### PR TITLE
[BIOMAGE-1929] - Remove infra hardcodings

### PR DESCRIPTION
**Pull request by aerlaut**
# Background
Removing hard coding from infra

#### Link to issue 
https://biomage.atlassian.net/browse/BIOMAGE-1929

#### Anything else the reviewers should know about the changes here

# Changes
### Code changes

# Definition of DONE
Your changes will be ready for merging after each of the steps below have been completed:

### Testing
- [ ] Unit tests written
- [ ] Tested locally with Inframock  (with latest production data downloaded with biomage experiment pull)
- [ ] Deployed to staging

To set up easy local testing with inframock, follow the instructions here: https://github.com/hms-dbmi-cellenics/inframock
To deploy to the staging environment, follow the instructions here: https://github.com/hms-dbmi-cellenics/biomage-utils

### Documentation updates
Is all relevant documentation updated to reflect the proposed changes in this PR?

- [ ] Relevant Github READMEs updated
- [ ] Relevant wiki pages created/updated

### Approvers
- [ ] Approved by a member of the core engineering team
- [ ] (UX changes) Approved by vickymorrison (this is her username, tag her if you need approval)

### Just before merging:
- [ ] After the PR is approved, the `unstage` script in here: https://github.com/hms-dbmi-cellenics/biomage-utils is executed. This script cleans up your deployment to staging

### Optional
- [ ] Photo of a cute animal attached to this PR

### CODE REVIEW

File: .github/workflows/deploy-infra.yaml [hunk 1]
```diff
@@ -206,26 +212,37 @@ jobs:
...
           if [ "${{ matrix.environment }}" = "staging" ]; then
-            DOMAIN_NAME="scp-staging.biomage.net"
+            DOMAIN_NAME=${{ secrets.DOMAIN_NAME_STAGING }}
           fi
 
-          AWS_REGION="${{ env.region }}"
-          AWS_ACCOUNT_ID="${{ steps.setup-aws.outputs.aws-account-id }}"
+          if [ -z "$DOMAIN_NAME" ]; then
+            echo "No domain name provided. Set DOMAIN_NAME_PRODUCTION or DOMAIN_NAME_STAGING in repository secrets."
+            exit 1
```
Comment (ivababukova) [inline .github/workflows/deploy-infra.yaml]: could you use this error message in line 59 too? It is better than what you have on line 59 at the moment

File: cf/post-register-lambda.yaml [hunk 2]
```diff
@@ -87,8 +87,11 @@ Resources:
     Type: 'AWS::SNS::Subscription'
     Properties:
       TopicArn: !Ref PostRegisterTopic
-      Endpoint: !If [isProd, "https://api.scp.biomage.net/v2/access/post-registration",
-                !Sub "https://api-default.scp-${Environment}.biomage.net/v2/access/post-registration"]
+      Endpoint: !Sub [ "${BaseUrl}/v2/access/post-registration", BaseUrl: !If [
```
Comment (ivababukova) [inline cf/post-register-lambda.yaml:90]: where is `BaseUrl` defined?

File: infra/cf-route53.yaml [hunk 3]
```diff
@@ -23,23 +23,25 @@ Parameters:
     Type: String
     Description: The ID of the hosted zone
 
+Outputs:
+  DomainName:
+    Value: !Ref DomainName
+    Export:
+      Name: !Sub "DomainName-${Environment}"
```
Comment (ivababukova) [inline infra/cf-route53.yaml:33]: And this output `DomainName` is used in the other CF files, is that right?

File: cf/sns.yaml [hunk 4]
```diff
@@ -22,24 +22,26 @@ Conditions:
...
     Properties:
       TopicArn: !Ref SNSTopic
-      Endpoint: !If [isProd, "https://api.scp.biomage.net/v1/pipelineResults",
-                !If [isDev, "http://host.docker.internal:3000/v1/pipelineResults",
-                !Sub "https://api-${SandboxID}.scp-${Environment}.biomage.net/v1/pipelineResults"]]
+      Endpoint: !Sub [ "${BaseUrl}/v1/pipelineResults", BaseUrl: !If [
+                        isDev, "http://host.docker.internal:3000", !If [
+                        isProd, !Sub ["https://api.${DomainName}", DomainName: !Join ["", [ Fn::ImportValue: !Sub "DomainName-${Environment}" ]]],
+                        !Sub ["https://api-${SandboxID}.${DomainName}", DomainName: !Join ["", [ Fn::ImportValue: !Sub "DomainName-${Environment}" ]]]
+                      ]]]
```
Comment (ivababukova) [inline cf/sns.yaml:44]: Instead of having this ifs every time you want to use `BaseUrl`, you can define a parameter at the top of the file and reuse it in this and the other places.

File: cf/userpoolclient.yaml [hunk 5]
```diff
@@ -27,8 +27,11 @@ Resources:
       UserPoolId:
         Fn::ImportValue: !Sub "biomage-user-pool-case-insensitive-${Environment}::UserPoolId"
       CallbackURLs:
-        - !If [isProd, "https://scp.biomage.net/data-management",
-          !Sub "https://ui-${SandboxID}.scp-${Environment}.biomage.net/data-management"]
+        - !Sub ["${BaseUrl}/data-management", BaseUrl:
+            !If [ isProd,
+              !Sub ["https://${DomainName}", DomainName: !Join ["", [ Fn::ImportValue: !Sub "DomainName-${Environment}" ]]],
+              !Sub ["https://ui-${SandboxID}.${DomainName}", DomainName: !Join ["", [ Fn::ImportValue: !Sub "DomainName-${Environment}" ]]]
+          ]]
```
Comment (ivababukova) [inline cf/userpoolclient.yaml:34]: same here -- better define a parameter for BaseUrl

File: infra/cf-route53.yaml (hunk 3 above)
Comment (aerlaut) [reply infra/cf-route53.yaml:33]: Yes, it is used in the CF files which require domain name. There are 2 stacks created from this, 1 for staging and 1 for production. Each output contains the staging or production domain name, as defined in Github secrets.

File: cf/sns.yaml (hunk 4 above)
Comment (aerlaut) [reply cf/sns.yaml:44]: I've tried this - according to the linter, Cloudformation does not allow the use of `Fn::ImportValue` in parameters or conditions

File: cf/userpoolclient.yaml (hunk 5 above)
Comment (aerlaut) [reply cf/userpoolclient.yaml:34]: I've tried this too. According to the linter, Cloudformation does not allow the use of `Fn::ImportValue` in parameters or conditions.

File: cf/sns.yaml (hunk 4 above)
Comment (aerlaut) [reply cf/sns.yaml:44]: - Can not use Fn::ImportValue in parameter ([ref](https://stackoverflow.com/q/60446398/1940886))
- Can not use Fn::ImportValue in conditions ([ref](https://stackoverflow.com/a/62076636/1940886))

File: .github/workflows/deploy-infra.yaml (hunk 1 above)
Comment (aerlaut) [reply .github/workflows/deploy-infra.yaml]: On 2nd thoughts, we don't actually need this check if we have the check on line 59. I'll check that the guarantee applies. If it is so, I'll remove this line.

File: cf/userpoolclient.yaml (hunk 5 above)
Comment (ivababukova) [reply cf/userpoolclient.yaml:34]: ah that's annoying

Comment (ivababukova) [reply cf/userpoolclient.yaml:34]: is there any other place (non parameters place) where we could define this BaseUrl variable?

Comment (aerlaut) [reply cf/userpoolclient.yaml:34]: I don't think there's an easy way to do this right now. CloudFormation allows for the creation and definition of [CustomResources](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/template-custom-resources.html), and according to [this SO post](https://serverfault.com/a/812040), with the help of a lambda, we can use this to make "variables". It sounds like a workaround though and might be too much for the scope that we have and functionality that we want.

File: .github/workflows/deploy-changed-cf.yaml [hunk 6]
```diff
@@ -1,5 +1,15 @@
...
+    inputs:
+      cluster:
+        type: choice
+        description: Select environment to deploy to
+        options:
+        - staging
+        - production
+        - staging and production
+        default: staging
   push:
```
Comment (ivababukova) [inline .github/workflows/deploy-changed-cf.yaml:13]: didn't you do these changes and merge them as part of a different PR?

Comment (aerlaut) [reply .github/workflows/deploy-changed-cf.yaml:13]: I added them here so that I can trigger this branch manually

File: .github/workflows/deploy-changed-cf.yaml [hunk 7]
```diff
@@ -16,49 +26,87 @@ env:
...
+          echo "[]" > empty.json
 
-      - id: check-number-of-cf-files
-        name: Check CloudFormation templates
-        run: |-
-          # Merge all modified and renamed files into one JSON array
-          #jq -s add ${HOME}/files_added.json ${HOME}/files_modified.json ${HOME}/files_renamed.json > ${HOME}/files_to_check.json
+          # Get name of changed files
+          # and concatenate into a JSON array
+          git diff --name-only --diff-filter=ACMRT remotes/origin/master ${{ github.sha }} | xargs -I 'value' jq --arg filename "value" '. |= . + [$filename]' empty.json | jq -s 'add' > ${HOME}/files.json
```
Comment (ivababukova) [inline .github/workflows/deploy-changed-cf.yaml:83]: why did you change this?

File: .github/workflows/deploy-changed-cf.yaml [hunk 8]
```diff
@@ -16,49 +26,87 @@ env:
...
     outputs:
       files: ${{ steps.check-number-of-cf-files.outputs.files }}
       num-files: ${{ steps.check-number-of-cf-files.outputs.num-files }}
     steps:
+      - id: checkout
+        name: Check out current branch source code
+        uses: actions/checkout@v2
+        with:
+          fetch-depth: 0
+
```
Comment (ivababukova) [inline .github/workflows/deploy-changed-cf.yaml:75]: why did you add this?

File: .github/workflows/deploy-changed-cf.yaml [hunk 9]
```diff
@@ -16,49 +26,87 @@ env:
...
           # Set as output the minified JSON.
           echo "::set-output name=files::$(jq -c . < ${HOME}/cf_files.json)"
           echo "$(jq -c . < ${HOME}/cf_files.json)"
           echo "::set-output name=num-files::$(jq '. | length' ${HOME}/cf_files.json)"
 
-  deploy-templates:
-    name: Deploy changed CloudFormation template
+  lint-templates:
+    name: Lint template files
```
Comment (ivababukova) [inline .github/workflows/deploy-changed-cf.yaml:103]: why did you change this?

File: .github/workflows/deploy-changed-cf.yaml [hunk 10]
```diff
@@ -113,8 +171,16 @@ jobs:
...
+      - id: get-oidc
+        if: ${{ contains(matrix.template, 'irsa-') }}
+        name: Get OIDC provider information for IRSA role
+        run: |-
+          OIDC_PROVIDER=$(aws eks describe-cluster --name "biomage-$CLUSTER_ENV" --query "cluster.identity.oidc.issuer" --output text | sed -e "s/^https:\/\///")
+          echo "::set-output name=oidc-provider::$OIDC_PROVIDER"
+
       - id: deploy-irsa-template
-        if: ${{ contains(matrix.template, 'irsa-') && github.ref == 'refs/heads/master' }}
+        if: ${{ contains(matrix.template, 'irsa-') }}
```
Comment (ivababukova) [inline .github/workflows/deploy-changed-cf.yaml:183]: why did you remove `github.ref == 'refs/heads/master'`?

File: .github/workflows/deploy-changed-cf.yaml [hunk 11]
```diff
@@ -204,8 +270,10 @@ jobs:
                 CREATE ROLE api_role WITH LOGIN;
                 CREATE ROLE dev_role WITH LOGIN;
                 GRANT USAGE ON SCHEMA public TO api_role;
+                GRANT USAGE, SELECT ON ALL SEQUENCES IN SCHEMA public to api_role;
                 GRANT dev_role TO ${AURORA_${{ steps.setup-rds-env.outputs.uppercase }}_USERNAME};
                 ALTER DEFAULT PRIVILEGES FOR USER dev_role IN SCHEMA public GRANT SELECT, INSERT, UPDATE, DELETE ON TABLES TO api_role;
+                ALTER DEFAULT PRIVILEGES FOR USER dev_role IN SCHEMA public GRANT USAGE, SELECT ON SEQUENCES TO api_role;
                 REVOKE dev_role FROM ${AURORA_${{ steps.setup-rds-env.outputs.uppercase }}_USERNAME};
```
Comment (ivababukova) [inline .github/workflows/deploy-changed-cf.yaml:277]: why do we need these changes?

File: README.md [hunk 12]
```diff
@@ -11,6 +11,33 @@ infrastructure is deployed *manually* by launching the action *Deploy Biomage in
 from the [actions](https://github.com/hms-dbmi-cellenics/iac/actions?query=workflow%3A%22Deploy+Biomage+infrastructure+on+AWS%22).
 The configuration for these infraustructure components are under `infra/`.
 
+#### Required secrets to deploy infrastructure
```
Comment (ivababukova) [inline README.md:14]: good idea to add this in the readme!

File: cf/bioinformatics-access-group.yaml [hunk 13]
```diff
@@ -35,7 +35,9 @@ Resources:
                   - 's3:PutObject'
                 Resource:
                   - !Sub "arn:aws:s3:::biomage-originals-${Environment}/*"
+                  - !Sub "arn:aws:s3:::biomage-originals-${Environment}-${AWS::AccountId}/*"
                   - !Sub "arn:aws:s3:::biomage-source-${Environment}/*"
+                  - !Sub "arn:aws:s3:::biomage-source-${Environment}-${AWS::AccountId}/*"
```
Comment (ivababukova) [inline cf/bioinformatics-access-group.yaml:40]: Are these policies a duplicate of the s3 policies in the developer-access-group?

File: .github/workflows/deploy-changed-cf.yaml (hunk 7 above)
Comment (aerlaut) [reply .github/workflows/deploy-changed-cf.yaml:83]: This is because the action `lots0logs/gh-action-get-changed-files@2.1.4` can only be used for `pull` and `push` actions, meanwhile we want to do this also for `workflow_dispatch` action. This was addressed here: https://github.com/hms-dbmi-cellenics/iac/pull/428#discussion_r896794017.

File: cf/bioinformatics-access-group.yaml (hunk 13 above)
Comment (aerlaut) [reply cf/bioinformatics-access-group.yaml:40]: No, these changes are made so that it's up to date with the master branch for this file.

File: .github/workflows/deploy-changed-cf.yaml (hunk 11 above)
Comment (aerlaut) [reply .github/workflows/deploy-changed-cf.yaml:277]: These changes are made to update this CF file up to date with master.

File: .github/workflows/deploy-changed-cf.yaml (hunk 10 above)
Comment (aerlaut) [reply .github/workflows/deploy-changed-cf.yaml:183]: It is already contained in the `if` statement at the job level.

File: .github/workflows/deploy-changed-cf.yaml (hunk 9 above)
Comment (aerlaut) [reply .github/workflows/deploy-changed-cf.yaml:103]: I split this job to be independent from the deploy job.

File: .github/workflows/deploy-infra.yaml [hunk 14]
```diff
@@ -54,12 +54,18 @@ jobs:
             ERROR=true
           fi
 
+          if [ -z "${{ secrets.PRIMARY_DOMAIN_NAME }}" ] || [ -z "${{ secrets.DOMAIN_NAME }}" ]
+          then
+            echo "No domain name provided. Set PRIMARY_DOMAIN_NAME and DOMAIN_NAME in repository secrets."
+            exit 1
+          fi
```
Comment (ivababukova) [inline .github/workflows/deploy-infra.yaml]: I would change the error message to be more accurate. Something like this:

PRIMARY_DOMAIN_NAME or DOMAIN_NAME is not set. Make sure that they are both set in repository secrets.

Or I would split the if into 2 ifs - one if for if PRIMARY_DOMAIN_NAME exists and one if for DOMAIN_NAME

File: README.md [hunk 15]
```diff
@@ -11,6 +11,34 @@ infrastructure is deployed *manually* by launching the action *Deploy Biomage in
...
+  The domain name that Cellenics will be deployed under (e.g. scp.biomage.net)
+
+- PRIMARY_DOMAIN_NAME
+
+  The primary domain name where Cellenics will be deployed under (e.g. biomage.net). This is the name of the domain
+  in the hosted zone.
+
+- DOMAIN_NAME_STAGING (optional, for Biomage use only)
+
+  The subdomain that the staging instance of Cellenics will be deployed (e.g. scp-staging.biomage.net)
```
Comment (ivababukova) [inline README.md]: replace "for Biomage use only" to "used only if the deployment is not on a production environment".

**Review by ivababukova — APPROVED** [review]
